### PR TITLE
feat(template): Add a flag to include canary analysis settings in the canary template. CLOUD-1116

### PR DESCRIPTION
**Pull request by th3morg**
## How has this been tested? How can a reviewer test these changes?

# Have your performed the following tests?

Please confirm you have done the following tests to ensure your pull request is ready to be merged..

- [x] If this change modifies the deployment flow have triggered a deployment using this branch (i.e. `build/dist/darwin_amd64/armory deploy start -f <path to file>`)?
- [x] If this change modifes the login flow have you tried logging in and out with this branch (i.e. `build/dist/darwin_amd64/armory login`) and verified the credentials work?
- [x] Have you verified the specific change made in this PR?

### CODE REVIEW

**Comment by th3morg** [comment]
Run `armory template kubernetes canary -f automated` to see the template. Option `-f manual` is also available which will show manual approval pause steps. `manual` is also the default when no flags are provided. To see the new options use the help menu `armory template kubernetes canary -h`

**Review by ivanphdz — DISMISSED** [review]
LGTM

File: cmd/template/canary.go [hunk 1]
```diff
@@ -69,11 +69,33 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+		options.features = append(options.features, "manual")
+	}
+	for _, feature := range options.features {
+		switch {
+		case feature == "manual":
+			stepsValuesNode.Content = append(stepsValuesNode.Content, pause, weight, pauseUA)
+		case feature == "automated":
+			// automated adds an analysis definition at the root level
+			analysisNode, analysisValuesNode := util.BuildMapNode("analysis", "Define queries and thresholds used for automated analysis")
+			defaultMetricProviderNameNode := util.BuildStringNode("defaultMetricProviderName", "prometheus-prod", "Optional. Name of the default provider to use for the queries. Configure in the Armory Cloud Console.")
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
			defaultMetricProviderNameNode := util.BuildStringNode("defaultMetricProviderName", "prometheus-prod", "Optional. Name of the default provider to use for the queries. Add providers in the Configuration UI.")
```

File: cmd/template/canary.go [hunk 2]
```diff
@@ -69,11 +69,33 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	if len(options.features) == 0 {
+		options.features = append(options.features, "manual")
+	}
+	for _, feature := range options.features {
+		switch {
+		case feature == "manual":
+			stepsValuesNode.Content = append(stepsValuesNode.Content, pause, weight, pauseUA)
+		case feature == "automated":
+			// automated adds an analysis definition at the root level
+			analysisNode, analysisValuesNode := util.BuildMapNode("analysis", "Define queries and thresholds used for automated analysis")
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
			analysisNode, analysisValuesNode := util.BuildMapNode("analysis", "Specify queries and thresholds used for automated analysis.")
```

The queries are actually defined in the analysis block, and they're picking them here, right?

At least in my head, defining == writing them out and specifying == picking

File: cmd/template/canary.go [hunk 3]
```diff
@@ -56,11 +58,9 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
 	pause.Content = append(pause.Content, pauseNode, pauseValuesNode)
 
-	// Weight root
-	weight := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
-	weightNode, weightValuesNode := util.BuildMapNode("setWeight", "")
-	weightValuesNode.Content = append(weightValuesNode.Content, util.BuildIntNode("weight", "33", "The percentage of pods that should be running the canary version for this step. Set it to an integer between 0 and 100, inclusive.")...)
-	weight.Content = append(weight.Content, weightNode, weightValuesNode)
+	// Weight nodes
+	weight := buildWeightStepNode("33", "The percentage of pods that should be running the canary version for this step. Set it to an integer between 0 and 100, inclusive.")
+	weight100 := buildWeightStepNode("100", "Setting weight to 100 is optional. Traffic will go to 100 after passing the final step.")
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
	weight100 := buildWeightStepNode("100", "Setting weight to 100 is optional. Traffic automatically goes to 100 after passing the final step.")
```

File: cmd/template/canary.go [hunk 4]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+func buildAnalysisQueryDefinitionNode(name string, metricProviderName string, upperLimit string, lowerLimit string, queryTemplate string) *yaml.Node {
+	query := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
+	query.Content = append(query.Content, util.BuildStringNode("name", name,
+		""	)...)
+	if len(metricProviderName) > 0 {
+		query.Content = append(query.Content, util.BuildStringNode("metricProviderName", metricProviderName,
+			"Optional. Override the defaultMetricProviderName specified in analysis.queries."	)...)
+	}
+	query.Content = append(query.Content, util.BuildIntNode("upperLimit", upperLimit,
+		"Optional when 'lowerLimit' is specified. If the metric exceeds this value, the automated analysis fails."	)...)
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
		"Optional when 'lowerLimit' is specified. If the metric exceeds this value, the automated analysis fails, causing the step to fail."	)...)
```

File: cmd/template/canary.go [hunk 5]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	query.Content = append(query.Content, util.BuildIntNode("lowerLimit", lowerLimit,
+		"Optional when 'upperLimit' is specified. If the metric goes below this value, the automated analysis fails."	)...)
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
		"Optional when 'upperLimit' is specified. If the metric goes below this value, the automated analysis fails, causing the step to fail."	)...)
```

File: cmd/template/canary.go [hunk 6]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	query.Content = append(query.Content, util.BuildStringNode("queryTemplate", queryTemplate,"")...)
+	return query
+}
+
+func buildAutomatedPauseStep() []*yaml.Node {
+	pauseUA := &yaml.Node{Kind: yaml.MappingNode, Tag: "!!map"}
+	pauseUANode, pauseUAValuesNode := util.BuildMapNode("analysis", "An analysis step pauses the deployment until analysis judgement runs are completed.")
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
	pauseUANode, pauseUAValuesNode := util.BuildMapNode("analysis", "An analysis step pauses the deployment until analysis judgement runs complete.")
```

File: cmd/template/canary.go [hunk 7]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	pauseUAValuesNode.Content = append(pauseUAValuesNode.Content, util.BuildIntNode("interval", "7",
+	"How long each sample of the query gets summarized over"	)...)
+	pauseUAValuesNode.Content = append(pauseUAValuesNode.Content, util.BuildStringNode("units", "seconds",
+		"The unit for the interval: 'seconds', 'minutes' or 'hours'"	)...)
+	pauseUAValuesNode.Content = append(pauseUAValuesNode.Content, util.BuildIntNode("numberOfJudgmentRuns", "1",
+	"How many times the queries are run.")...)
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
	"How many times the queries run.")...)
```

File: cmd/template/canary.go [hunk 8]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	queriesNode, queriesValuesNode := util.BuildSequenceNode("queries", "rollBackMode: manual # Optional. Defaults to 'automatic' if omitted. Uncomment to require a manual review before rolling back if automated analysis detects an issue.\n" +
+		"rollForwardMode: manual # Optional. Defaults to 'automatic' if omitted. Uncomment to require a manual review before continuing deployment if automated analysis determines the application is healthy.")
```
Comment (317brian) [inline cmd/template/canary.go]:
```suggestion
		"rollForwardMode: manual # Optional. Defaults to 'automatic' if omitted. Uncomment to require a manual review before continuing deployment if automated analysis determines the app is healthy.")
```

File: cmd/template/canary.go [hunk 9]
```diff
@@ -86,3 +108,63 @@ func canary(cmd *cobra.Command, options *templateCanaryOptions, args []string) e
...
+	queriesValuesNode.Content = append(queriesValuesNode.Content,
+		&yaml.Node{Kind: yaml.ScalarNode, Tag: "!!str", Value: "containerCPUSeconds",
+			HeadComment: "Specify a list of queries to run. Reference them by the name you assign in analysis.queries."},
```
Comment (317brian) [inline cmd/template/canary.go:156]:
```suggestion
			HeadComment: "Define a list of queries to run. Reference them by the name you assign in analysis.queries."},
```

**Review by 317brian — DISMISSED** [review]
Some minor nitpicking on my part, but otherwise LGTM

File: cmd/template/canary.go (hunk 2 above)
Comment (th3morg) [reply cmd/template/canary.go]: This is actually the analysis block where they are defining the queries.

**Review by 317brian — APPROVED** [review]
🥳

**Pull request merged**